File: order/src/main/java/com/trifail/order/config/db/DataSourceConfig.java
package com.trifail.order.config.db;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.jdbc.DataSourceBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.core.io.ClassPathResource;
import org.springframework.dao.annotation.PersistenceExceptionTranslationPostProcessor;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.jdbc.datasource.init.DataSourceInitializer;
import org.springframework.jdbc.datasource.init.ResourceDatabasePopulator;
import org.springframework.orm.jpa.JpaTransactionManager;
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import javax.persistence.EntityManagerFactory;
import javax.sql.DataSource;
import java.util.Properties;

@Configuration
@EnableTransactionManagement
@EnableJpaRepositories(basePackages = "com.trifail.order.repository")
public class DataSourceConfig {

    @Autowired
    private ExtraSrouceConfig extraSrouceConfig;

    @Bean
    @Primary
    @ConfigurationProperties(prefix = "first.datasource")
    public DataSource mysqlDataSource() {
        return DataSourceBuilder.create()
                .url(extraSrouceConfig.url)
                .driverClassName(extraSrouceConfig.driver)
                .username(extraSrouceConfig.username)
                .password(extraSrouceConfig.password)
                .build();
    }

    /**
     * 事务管理著接口
     */
    @Bean
    public PlatformTransactionManager transactionManager(EntityManagerFactory emf) {
        JpaTransactionManager transactionManager = new JpaTransactionManager();
        transactionManager.setEntityManagerFactory(emf);
        return transactionManager;
    }

    @Bean
    public PersistenceExceptionTranslationPostProcessor exceptionTranslation() {
        return new PersistenceExceptionTranslationPostProcessor();
    }

    /**
     * EntityManager
     */
    @Bean
    public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
        LocalContainerEntityManagerFactoryBean factory = new LocalContainerEntityManagerFactoryBean();
        factory.setDataSource(mysqlDataSource());
        factory.setPackagesToScan("com.trifail.order.model");
        //hibernate jpa
        factory.setJpaVendorAdapter(new HibernateJpaVendorAdapter());
        //customized proeprties
        Properties jpaProperties = new Properties();
        jpaProperties.put("hibernate.hbm2ddl.auto", extraSrouceConfig.ddlAuto);
        jpaProperties.put("hibernate.show-sql", extraSrouceConfig.showSql);
        jpaProperties.put("hibernate.dialect", extraSrouceConfig.dialect);
        jpaProperties.put("hibernate.format_sql", extraSrouceConfig.formatable);
        factory.setJpaProperties(jpaProperties);
        return factory;
    }


    /**
     * 数据库是否初始化
     */
    @Bean
    public DataSourceInitializer dataSourceInitializer() {
        DataSourceInitializer dsInitializer = new DataSourceInitializer();
        dsInitializer.setDataSource(mysqlDataSource());
        ResourceDatabasePopulator dbPopulator = new ResourceDatabasePopulator();
        dbPopulator.addScript(new ClassPathResource("order_initialization.sql"));
        dsInitializer.setDatabasePopulator(dbPopulator);
        dsInitializer.setEnabled(extraSrouceConfig.initialized);
        return dsInitializer;
    }
}

File: store/src/main/kotlin/com/trifail/store/config/DataSourceConfig.kt
package com.trifail.store.config

import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.jdbc.DataSourceBuilder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.core.io.ClassPathResource
import org.springframework.dao.annotation.PersistenceExceptionTranslationPostProcessor
import org.springframework.data.jpa.repository.config.EnableJpaRepositories
import org.springframework.jdbc.datasource.init.DataSourceInitializer
import org.springframework.jdbc.datasource.init.ResourceDatabasePopulator
import org.springframework.orm.jpa.JpaTransactionManager
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter
import org.springframework.transaction.PlatformTransactionManager
import java.util.*
import javax.persistence.EntityManagerFactory
import javax.sql.DataSource

/**
 * copy from store
 */
@Configuration
@EnableJpaRepositories(basePackages = ["com.trifail.store.repository"])
class DataSourceConfig{

    @Value("\${store.mysql.datasource.url}")
    private val url: String? = "jdbc:mysql://localhost:3306/store?characterEncoding=utf8&useSSL=false"
    @Value("\${store.mysql.datasource.driver}")
    private val driver: String? = "com.mysql.cj.jdbc.Driver"
    @Value("\${store.mysql.datasource.username}")
    private val username: String? = "root"
    @Value("\${store.mysql.datasource.password}")
    private val password: String? = "123456"
    @Value("\${store.mysql.datasource.withInitialization}")
    private val initialized: Boolean? = false
    @Value("\${store.jpa.show.sql}")
    private val showSql: Boolean? = false
    @Value("\${store.jpa.ddl.auto}")
    private val ddlAuto: String? = "UPDATE"
    @Value("\${store.jpa.dialect}")
    private val dialect: String? = "org.hibernate.dialect.MySQL8Dialect"
    @Value("\${store.jpa.format_sql}")
    private val formatable: Boolean? = true


    @Bean
    @Primary
    @ConfigurationProperties(prefix = "first.datasource")
    fun mysqlDataSource(): DataSource {
        val build = DataSourceBuilder.create()
                .url(url)
                .driverClassName(driver)
                .username(username)
                .password(password)
                .build()
        return build;
    }

    @Bean
    fun transactionManager(emf: EntityManagerFactory): PlatformTransactionManager {
        val transactionManager = JpaTransactionManager()
        transactionManager.entityManagerFactory = emf
        return transactionManager
    }

    @Bean
    fun exceptionTranslation(): PersistenceExceptionTranslationPostProcessor {
        return PersistenceExceptionTranslationPostProcessor()
    }

    /**
     * EntityManager
     */
    @Bean
    fun entityManagerFactory(): LocalContainerEntityManagerFactoryBean {
        val factory = LocalContainerEntityManagerFactoryBean()
        factory.dataSource = mysqlDataSource()
        factory.setPackagesToScan("com.trifail.store.model")
        //hibernate jpa
        factory.jpaVendorAdapter = HibernateJpaVendorAdapter()
        //customized proeprties
        val jpaProperties = Properties()
        jpaProperties["hibernate.hbm2ddl.auto"] = ddlAuto!!
        jpaProperties["hibernate.show-sql"] = showSql!!
        jpaProperties["hibernate.dialect"] = dialect!!
        jpaProperties["hibernate.format_sql"] = formatable!!
        factory.setJpaProperties(jpaProperties)
        return factory
    }


    /**
     * 数据库是否初始化
     */
    @Bean
    fun dataSourceInitializer(): DataSourceInitializer {
        val dsInitializer = DataSourceInitializer()
        dsInitializer.setDataSource(mysqlDataSource())
        val dbPopulator = ResourceDatabasePopulator()
        dbPopulator.addScript(ClassPathResource("store_initialization.sql"))
        dsInitializer.setDatabasePopulator(dbPopulator)
        dsInitializer.setEnabled(initialized!!)
        return dsInitializer
    }
}